pcdsdevices/signal.py

```diff
@@ -157,55 +194,222 @@ def _calc_readback(self):
+            "sent_value_callback": False,
+        }
+
+        yield state_info
+
+        # Now that the caller has updated state, check to see if our
+        # connectivity status has changed.  Update the mutable return
+        # value at this point.
+        is_connected = self.connected
+        state_info["is_connected"] = is_connected
+        if was_connected != is_connected:
+            self._metadata["connected"] = is_connected
+            self._run_metadata_callbacks()
+            state_info["sent_md_callback"] = True
+
+        if not was_connected and is_connected:
+            # disconnected -> connected; give a proper value callback.
+            # "connected" indicates that:
+            # 1. All underlying signals are connected
+            # 2. All underlying signals have a value cached
+            with self._lock:
+                old_value = self._readback
+                self._readback = self._calc_readback()
+                self._run_subs(
+                    sub_type=self.SUB_VALUE,
+                    obj=self,
+                    value=self._readback,
+                    old_value=old_value,
+                )
+                state_info["sent_value_callback"] = True
+
+    def add_signal_by_attr_name(self, name: str) -> Signal:
+        """
+        Add a signal from which to aggregate information.
+
+        This must be called before any subscriptions are made to the signal.
+        Duplicate signals will not be re-added.
+
+        Parameters
+        ----------
+        name : str
+            The attribute name of the signal, relative to the parent device.
+
+        Returns
+        -------
+        sig : ophyd.Signal
+            The signal referenced by the attribute name.
+
+        Raises
+        ------
+        RuntimError
+            If called after .subscribe() or used without a parent Device.
+        """
+        if self._has_subscribed:
+            raise RuntimeError(
+                "Cannot add signals to an AggregateSignal after it has been "
+                "subscribed to."
+            )
+
+        sig = self.parent
+
+        if sig is None:
+            raise RuntimeError(
+                "Cannot use an AggregateSignal with attribute names outside "
+                "of a Device/Component hierarchy."
+            )
+
+        for part in name.split('.'):
+            sig = getattr(sig, part)
+
+        # Add if not yet there; but do not subscribe just yet.
+        self._signals[sig] = _AggregateSignalState(signal=sig)
+        return sig
+
+    def destroy(self):
+        for signal, info in self._signals.items():
+            if info.value_cbid is not None:
+                signal.unsubscribe(info.value_cbid)
+                info.value_cbid = None
+
+            if info.meta_cbid is not None:
+                signal.unsubscribe(info.meta_cbid)
+                info.meta_cbid = None
+
+        return super().destroy()
+
 
 class PVStateSignal(AggregateSignal):
```

Comment:
I think the refactor here maintains the original behavior wrt values, I might opt to find and test a device or two though.

Reply:
Given how we use these in nearly everything, full testing would be strongly recommended...